Store username with messages, auto-tag, and UI display

Tigger2014 opened the pull request:
This PR enhances Mailpit by storing the SMTP/HTTP-authenticated username with each message, making it available for tagging and display in the UI.
## Key changes

- **Database & Backend**
  - Added a `Username` column to the `mailbox` table (with migration).
  - Updated message storage to save the username (if present) with each message.
  - Updated `DBMailSummary`, `MessageSummary`, and `Message` structs to include a `Username` field.

- **Tagging**
  - Added support for auto-tagging messages with the username if the `--tags-username` / `MP_TAGS_USERNAME` flag is set.

- **UI**
  - Updated the web UI (`Message.vue`) to display the username in the message header if present.

- **Testing & Debugging**
  - Updated and fixed Go tests and helpers for the new logic, including username auto-tagging.

- **Schema**
  - Added migration for the `Username` column (`1.27.0.sql`).

## Motivation

- Enables tracking and filtering of messages by user.

## Notes

- The initial schema and migrations have been reviewed to ensure a fresh DB is always correct.
- All tests pass after these changes.

axllent commented:
Hey @Tigger2014 , this is pretty awesome! I was going to suggest referring to it as the "smtp user", however I then realised that you also were also catering for the HTTP send API too, so never mind.... username makes sense :)

There is a fair bit of additional overhead/changes as a direct result of storing the username as a separate field in the database. I am wondering if it wouldn't have been a better idea to store it instead as a new key/value in the `DBMailSummary`? Maybe you did it this way for a reason, what are your thoughts?

Tigger2014 commented:
I think I did attempt to do it that way, maybe I've missed some logic here.

I added the username to `DBMailSummary`, which is populated upon storing the message.
However, `GetMessage` does not read DBMailSummary from what I can see and instead gets all info from the `mailbox_data` table.  So I then make an additional lookup to get the user from the mailbox table when retrieving a message so it can be appended to the json response.

Let me know if I've misunderstood something and I would be happier to refactor.

axllent commented:
I think the following should work nicely, let me try explain:

1. Rename the `DBMailSummary struct` to `Metadata struct` (after all, it's not a summary of the email, it just includes some  metadata which is used for the `storage.List()` and search filtering).
2. Store the Username in the metadata, remove the new column you added.
3. Create a standalone function in storage to return the populated metadata struct from a message (`func GetMetadata(id string) (Metadata, error){}` - this function can then be called in the `storage.GetMessage()` to easily load the metadata object, and therefore get the Username you need.
4. Update the reindex subcommand to generate the initial metadata object with `storage.GetMetadata()`  (from the existing message) before overwriting the From, To, Cc, Bcc and ReplyTo values - this means we wouldn't loose the Username and any other potential fields which may be added in the future.

Does this make sense to you? If you need any help or advice then please let me know. No rush from me, I'm not sure how much time (if any) I'll have during the week, and the earliest I'll release a new version would be on the weekend anyway.

Thanks again Ben!

Tigger2014 commented:
@axllent Thanks for the detailed feedback, It made sense to me.

I think I've addressed your concerns, no rush to review, this is not an urgently needed thing for me either :)

Please do let me know if you have any other suggestions or if I've misunderstood

axllent commented:
Great work @Tigger2014 ! I have now made a few cleanups and minor changes, and merged this into develop. Unfortunately something went wrong while merging the commits into one commit (I think because I removed the references to the merged commits), which unfortunately removed your details from the final merged commit (so it looks like I did all the work). I'm so sorry, I was definitely not trying to take any credit for your work, I was just trying to keep the merge commit clean. My mistake. I'll definitely mention you in the release notes!

Would you please be so kind to update the mailpit-website documentation for me with the new flags & environment details, if possible? Sorry, I'm really short on time at the moment with a tonne of other things to do (work).
